metaseq/distributed/rendezvous.py

```diff
@@ -70,4 +70,52 @@ def _error(msg):
     raise RuntimeError("Unable to perform re-rendezvous using tcpr:// method")
 
 
+STORE_BASED_BARRIER_MARKER = "store_based_barrier_key:"
+
+
+class ComplicitStore(Store):
+    def __init__(self, store: Store, world_size: int):
+        super().__init__()
+        self.store = store
+        self.world_size = world_size
+
+    def set(self, *args, **kwargs):
+        return self.store.set(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        return self.store.get(*args, **kwargs)
+
+    def add(self, key: str, *args, **kwargs):
+        # The marker isn't always a prefix: it's sometimes prefixed with default_pg/.
+        if STORE_BASED_BARRIER_MARKER in key:
+            return self.world_size
```

Comment:
I see, specifically if it's the initialization, we just pretend it's done.

I noticed this part of the docs on TCPStore:

https://pytorch.org/docs/stable/distributed.html#torch.distributed.TCPStore

Namely, on construction it has a `wait_for_worker=True` default parameter, which we are not currently setting. Maybe we can get the same effect by just turning this off?

Reply:
I see, there's actually two barriers.

One in C++, for TCP store:

https://github.com/pytorch/pytorch/blob/c18da597e0bb1c1aecc97c77a73fed1849057fa4/torch/csrc/distributed/c10d/TCPStore.cpp#L1012-L1050

And one in the process group initialization:
https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L851-L863

And then a third in our own code:
https://github.com/facebookresearch/metaseq/blob/2e882aafef307b200728759dc7e0544622bf4b8c/metaseq/distributed/utils.py#L224

So maybe we can combine both TCPStore(wait_for_worker=False) and this and get even better overlap?

Reply:
@stephenroller what do you want to combine with TCPStore, once again?

As far as I understand, you want to just add `wait_for_workers=False` in _create_c10d_store (https://github.com/facebookresearch/metaseq/blob/2e882aafef307b200728759dc7e0544622bf4b8c/metaseq/distributed/rendezvous.py#L25) and continue using tcpr?  

It looks to me like that should work. @lw what do you think?

Reply:
I think @stephenroller meant to do both this PR and pass the `wait_for_worker=False` flag to the TCPStore constructor, since they control two different barriers (both of which could be a problem).

I believe the `wait_for_workers=False` flag will have limited effect, because it only controls one barrier which happens in the constructor of the store, and the store is only constructed once! (The sub-PGs re-use the same store as the root PG, they don't create a new one). I do not know why that barrier was needed and what safety implications come with removing it, I'd suggest looking into that.

I'd personally recommend removing each of these barriers in a separate commit, to test their performance and correctness impact separately.

Reply:
+1 to separating the proposed wait_for_worker=False into an entirely separate commit.